### PR TITLE
V3 parity spec for ChallengeOktaVerifyPush_spec

### PR DESCRIPTION
**Pull request by vijetmahabaleshwar-okta**
## PR Checklist

- [ ] Have you verified the basic functionality for this change?
- [ ] Did you add tests, as appropriate, following our [Automated Test guidelines](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [ ] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [ ] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-509011](https://oktainc.atlassian.net/browse/OKTA-509011)

### CODE REVIEW

File: test/testcafe/spec/ChallengeOktaVerifyPush_spec.js [hunk 1]
```diff
@@ -89,61 +100,72 @@ test
...
     await challengeOktaVerifyPushPageObject.clickAutoChallengeCheckbox();
-    await t.expect(checkboxLabel.hasClass('checked')).notOk();
+    await t.expect(checkbox.checked).notOk();
   });
-
+  
 test
   .requestHooks(logger, pushSuccessMock)('challenge okta verify push request', async t => {
-    await setup(t);
-    await checkA11y(t);
```
Comment (glenfannin-okta) [inline test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:146]: Were these intended to be removed?

File: test/testcafe/spec/ChallengeOktaVerifyPush_spec.js [hunk 2]
```diff
@@ -167,7 +189,6 @@ test
 test
   .requestHooks(pushWaitMock)('Warning callout appears after 30 seconds', async t => {
     const challengeOktaVerifyPushPageObject = await setup(t);
-    await checkA11y(t);
```
Comment (glenfannin-okta) [inline test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:170]: Was this intentional?

File: test/testcafe/spec/ChallengeOktaVerifyPush_spec.js [hunk 3]
```diff
@@ -89,61 +100,72 @@ test
...
-    await t.expect(answerRequestBody).eql({
-      autoChallenge: true,
+    await t.expect(answerRequestBody).contains({
       stateHandle: '02PVkP3FJyDnqUKkkxIZhxbsx7a2S-hC1JxIE6AXzp',
     });
+    if (userVariables.v3) {
+      await t.expect(answerRequestBody).contains({
+        autoChallenge: true,
+      });
+    }
```
Comment (glenfannin-okta) [inline test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:156]: Before this change, it looks like v2 also expects autoChallenge to be set to true, was this change intended?

File: test/testcafe/spec/ChallengeOktaVerifyPush_spec.js [hunk 4]
```diff
@@ -61,24 +70,26 @@ test
...
     await t.expect(await challengeOktaVerifyPushPageObject.signoutLinkExists()).ok();
     await t.expect(challengeOktaVerifyPushPageObject.getSignoutLinkText()).eql('Back to sign in');
   });
 
-test
+// Disabled in v3 - https://oktainc.atlassian.net/browse/OKTA-594821
+test.meta('v3', false)
   .requestHooks(pushAutoChallengeMock)('challenge ov push screen has right labels and a checkbox', async t => {
     const challengeOktaVerifyPushPageObject = await setup(t);
-    await checkA11y(t);
```
Comment (glenfannin-okta) [inline test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:81]: Was this intentional?

File: test/testcafe/spec/ChallengeOktaVerifyPush_spec.js [hunk 5]
```diff
@@ -37,18 +45,19 @@ const pushWaitAutoChallengeMock = RequestMock()
...
   await challengeOktaVerifyPushPageObject.navigateToPage();
+  await t.expect(challengeOktaVerifyPushPageObject.formExists()).eql(true);
   return challengeOktaVerifyPushPageObject;
 }
 
 test
-  .requestHooks(pushSuccessMock)('challenge ov push screen has right labels and logo', async t => {
+  .requestHooks(pushPollMock)('challenge ov push screen has right labels and logo', async t => {
     const challengeOktaVerifyPushPageObject = await setup(t);
-    await checkA11y(t);
```
Comment (glenfannin-okta) [inline test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:51]: Was this intentional?

File: test/testcafe/spec/ChallengeOktaVerifyPush_spec.js [hunk 6]
```diff
@@ -89,61 +100,72 @@ test
...
+    await t.expect(await challengeOktaVerifyPushPageObject.verifyWithSomethingElseLinkExists()).ok();
     await t.expect(await challengeOktaVerifyPushPageObject.signoutLinkExists()).ok();
     await t.expect(challengeOktaVerifyPushPageObject.getSignoutLinkText()).eql('Back to sign in');
   });
 
+// V3 - Polling fails with AssertionError: expected 8 to deeply equal 1
 test
   .requestHooks(logger, pushWaitAutoChallengeMock)('should call polling API and checkbox should be clickable after polling started', async t => {
     const challengeOktaVerifyPushPageObject = await setup(t);
-    await checkA11y(t);
```
Comment (glenfannin-okta) [inline test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:115]: Intentional?

File: test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js [hunk 7]
```diff
@@ -1,86 +1,102 @@
...
-  getFormTitleWithError() {
-    const titlePosition = userVariables.v3 ? 1 : 0;
-    return this.form.getNthTitle(titlePosition);
+    return this.form.getElement(FORM_INFOBOX_ERROR_TITLE);
   }
 
   getWarningBox() {
     if (userVariables.v3) {
-      return this.form.getAlertBox();
+      return this.form.getElement(INFOBOX_WARNING_V3);
```
Comment (glenfannin-okta) [inline test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js:69]: Was there a reason that `getAlertBox()` did not work? It seems like it did the same thing as this (getting the box by the alert role)?

File: test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js [hunk 8]
```diff
@@ -1,86 +1,102 @@
...
+  async isPushButtonDisabled() {
+    const pushBtnClass = await this.getPushButton().getAttribute('class');
+    return pushBtnClass.includes('disabled');
+  }
+
   getA11ySpan() {
     return this.form.getElement('.accessibility-text');
   }
 
   getResendPushButton() {
```
Comment (leemchale-okta) [inline test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js:40]: is this method still used anywhere?

File: test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js [hunk 9]
```diff
@@ -1,86 +1,102 @@
...
-    }
     return this.form.clickSaveButton();
   }
 
-  clickResendNumberChallenge() {
+  async clickResendNumberChallenge() {
+    if (userVariables.v3) {
+      return this.form.clickElement('a');
+    }
     return this.form.clickElement(RESEND_NUMBER_CHALLENGE_BUTTON);
```
Comment (leemchale-okta) [inline test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js:52]: would `this.t.click(this.form.getLink('resend the push notification'));` work here instead?  To make it a little more semantic

File: test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js (hunk 7 above)
Comment (leemchale-okta) [reply test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js:69]: was it because the warning box and the error box share the same role of `alert` ?

Comment (leemchale-okta) [reply test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js:69]: I think `getAlertBox()` and `getErrorBox()` both look for an element with a role of `alert`

File: test/testcafe/spec/ChallengeOktaVerifyPush_spec.js (hunk 5 above)
Comment (vijetmahabaleshwar-okta) [reply test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:51]: Unintentional. Adding it back.

File: test/testcafe/spec/ChallengeOktaVerifyPush_spec.js (hunk 4 above)
Comment (vijetmahabaleshwar-okta) [reply test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:81]: Unintentional. Adding it back.

File: test/testcafe/spec/ChallengeOktaVerifyPush_spec.js (hunk 6 above)
Comment (vijetmahabaleshwar-okta) [reply test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:115]: Unintentional. Adding it back.

File: test/testcafe/spec/ChallengeOktaVerifyPush_spec.js (hunk 1 above)
Comment (vijetmahabaleshwar-okta) [reply test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:146]: Unintentional. Adding it back.

File: test/testcafe/spec/ChallengeOktaVerifyPush_spec.js (hunk 2 above)
Comment (vijetmahabaleshwar-okta) [reply test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:170]: Unintentional. Adding it back.

File: test/testcafe/spec/ChallengeOktaVerifyPush_spec.js (hunk 3 above)
Comment (vijetmahabaleshwar-okta) [reply test/testcafe/spec/ChallengeOktaVerifyPush_spec.js]: This was unintentional. Changed it back.

File: test/testcafe/spec/ChallengeOktaVerifyPush_spec.js [hunk 10]
```diff
@@ -89,61 +103,76 @@ test
...
-
+  
 test
   .requestHooks(logger, pushSuccessMock)('challenge okta verify push request', async t => {
-    await setup(t);
-    await checkA11y(t);
+    const challengeOktaVerifyPushPageObject = new ChallengeOktaVerifyPushPageObject(t);
+    await challengeOktaVerifyPushPageObject.navigateToPage();
+    if (!userVariables.v3) {
+      await checkA11y(t);
```
Comment (vijetmahabaleshwar-okta) [inline test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:174]: Fails in v3 with this error - 
```
         56 |const axeCheck = (t, context, options = { rules: {} }) => {
         57 |    try {
         58 |        // skipping the "document-title" rule as there is an issue with testcafe
         59 |        // being unable to find the title of a page inside the <head> tag.
         60 |        options.rules["document-title"] = { enabled: false };
       > 61 |        return runAxe.with({ boundTestRun: t })(context, options);
         62 |    } catch (error) {
         63 |        return Promise.resolve({ error });
         64 |    }
         65 |};
         66 |

         at axeCheck (/Users/vijetmahabaleshwar/Downloads/temp/okta-signin-widget/node_modules/@testcafe-community/axe/index.js:61:48)
         at <anonymous> (/Users/vijetmahabaleshwar/Downloads/temp/okta-signin-widget/test/testcafe/framework/a11y.js:35:29)
         at asyncGeneratorStep (/Users/vijetmahabaleshwar/Downloads/temp/okta-signin-widget/test/testcafe/framework/a11y.js:1:225)
         at _next (/Users/vijetmahabaleshwar/Downloads/temp/okta-signin-widget/test/testcafe/framework/a11y.js:1:563)
         at <anonymous> (/Users/vijetmahabaleshwar/Downloads/temp/okta-signin-widget/test/testcafe/framework/a11y.js:1:728)
         at <anonymous> (/Users/vijetmahabaleshwar/Downloads/temp/okta-signin-widget/test/testcafe/framework/a11y.js:1:469)
         at checkA11y (/Users/vijetmahabaleshwar/Downloads/temp/okta-signin-widget/test/testcafe/framework/a11y.js:29:23)
         at <anonymous> (/Users/vijetmahabaleshwar/Downloads/temp/okta-signin-widget/test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:173:11)
         at asyncGeneratorStep (/Users/vijetmahabaleshwar/Downloads/temp/okta-signin-widget/test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:9:391)
         at _next (/Users/vijetmahabaleshwar/Downloads/temp/okta-signin-widget/test/testcafe/spec/ChallengeOktaVerifyPush_spec.js:9:729)
 ```

File: test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js (hunk 7 above)
Comment (vijetmahabaleshwar-okta) [reply test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js]: getAlertBox()  works just fine. Changed it back.

File: test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js [hunk 11]
```diff
@@ -1,86 +1,99 @@
...
   getErrorTitle() {
-    if (userVariables.v3) {
-      return this.form.getNthTitle(0);
-    }
-    return this.form.getElement(FORM_INFOBOX_ERROR_TITLE).innerText;
-  }
-
-  getFormTitleWithError() {
-    const titlePosition = userVariables.v3 ? 1 : 0;
-    return this.form.getNthTitle(titlePosition);
```
Comment (glenfannin-okta) [inline test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js:63]: I think removing these will result in failures in other tests that use it: 

See the failed specs: 

```
challenge okta verify resend push with uv enable biometrics message
2023-04-11 21:24:39 
2023-04-11 21:24:39    1) TypeError: challengeOktaVerifyPushPageObject.getFormTitleWithError is not a function
```

File: test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js [hunk 12]
```diff
@@ -1,86 +1,99 @@
...
-    if (userVariables.v3) {
-      return this.form.getButton(RESEND_PUSH_NOTIFICATION_TEXT);
-    }
     return this.form.getElement('.button-primary');
   }
 
   clickResendPushButton() {
-    if (userVariables.v3) {
-      return this.form.clickSaveButton(RESEND_PUSH_NOTIFICATION_TEXT);
-    }
```
Comment (glenfannin-okta) [inline test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js:38]: I think this change will impact this spec: `Challenge Okta Verify Push Resend`

File: test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js [hunk 13]
```diff
@@ -1,86 +1,99 @@
...
-
     return this.form.getElement(FORM_INFOBOX_WARNING);
   }
 
   async autoChallengeInputExists() {
     return this.form.elementExist(AUTO_CHALLENGE_CHECKBOX_SELECTOR);
   }
 
-  getAutoChallengeCheckboxLabel() {
-    return this.form.getElement(AUTO_CHALLENGE_CHECKBOX_LABEL_SELECTOR);
```
Comment (glenfannin-okta) [inline test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js:79]: Changing this will make this spec fail: 
`Challenge Authenticator Okta Verify Push Only With Auto Challenge Form`
`should load view with a button and a checkbox when push is the only method type and has auto challenge schema`

**Review by glenfannin-okta — COMMENTED** [review]
The core changes look good but I think some of the changes to the existing functions in the page object are causing other specs to fail. Once those are fixed/reverted we can approve to merge this 🙏

File: test/testcafe/spec/ChallengeOktaVerifyPush_spec.js (hunk 10 above)
Comment (glenfannin-okta) [reply test/testcafe/spec/ChallengeOktaVerifyPush_spec.js]: Resolved this with the latest commit.

**Review by glenfannin-okta — APPROVED** [review]
